shared/challenge.go

```diff
@@ -4,10 +4,19 @@ import "github.com/spacemeshos/sha256-simd"
 
 type Challenge []byte
 
-var ZeroChallenge = make(Challenge, 0)
+var (
+	ZeroChallenge = make(Challenge, 0)
+	message []byte
+)
 
+// ⚠️ This method is NOT thread-safe
 func (ch Challenge) GetSha256Parent(lChild, rChild []byte) []byte {
-	children := append(lChild, rChild...)
-	res := sha256.Sum256(append(ch, children...))
+	if len(message) != len(ch)+len(lChild)+len(rChild) {
```

Comment:
We can work with `message` even if it's larger, no?
But since we don't expect this use-case here, it's not a must.

Reply:
Not bad! I've benchmarked the following and it's slightly faster:
```golang
func combineGlobal(x, y, z []byte) []byte {
	l := len(x) + len(y) + len(z)
	if len(m) < l {
		m = make([]byte, l)
	}
	copy(m, x)
	copy(m[len(x):], y)
	copy(m[len(x)+len(y):], z)
	return m[:l]
	// 11.7 ns/op
}
```

I'm pretty sure the compiler takes care of that `l` and it doesn't require an allocation. Will make sure before merging. If I use a global for `l` it's actually a little slower.